instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.instrumenter.async;
```

Comment:
Does this need to be in instrumentation API? It seems to only really be needed by one instrumentation, the annotations Instrumentation, isn't it?

Reply:
And springboot autoconfigure

Reply:
Yep, we need to keep it in some shared place.

> I did suggest at that time that maybe this was a little too specialized for instrumentation-api and maybe it would make sense to have a separate shared module for AOP concerns?

Hmm, I wonder about that - I fear that introducing yet another module may only increase the cognitive load of implementing an instrumentation.

@anuraaga WDYT about marking the whole package as `@UnstableApi`?

Reply:
Cognitive load is precisely why we might want to find some split for it. I think async return value instrumentation is far less common than others so having a package in our instrumentation API could be too discoverable for our generally expected user. Just because a server framework is async, returning CompletableFuture, doesn't mean it needs to care about this package but someone might think so just by the word async, I would have I think.

Marking as UnstableApi seems like a good idea, can we maybe name the package annotationsasyncsupport or something similar to tie it closer to the instrumentation?

Reply:
> Hmm, I wonder about that - I fear that introducing yet another module may only increase the cognitive load of implementing an instrumentation.

I guess it depends on what instrumentation we expect to rely on these classes.  If it's just the AOP instrumentation (`opentelemetry-annotations-1.0` and `spring-boot-autoconfigure`) as well as the different reactive frameworks (`guava`,`rxjava2`, `rxjava3` and `reactor`, with maybe a couple more being added over time), I don't think it's too problematic to spit it into a separate module as it's use will be fairly specialized.

However, if it's expected that instrumentation for various reactive client libraries (e.g. `lettuce`, `spring-webflux-5.0`) might also want to reuse the strategies, then it probably makes more sense for them to live in `instrumentation-api`.  That said I don't think the strategies are geared towards that use, especially now that some of them are configuration driven and there's an expectation that they are registered with the reactive framework they represent.  Although nothing would stop that instrumentation module from creating its own instance of the strategy rather than bouncing through `AsyncEndStrategies`.

Reply:
> having a package in our instrumentation API could be too discoverable for our generally expected user. Just because a server framework is async, returning CompletableFuture, doesn't mean it needs to care about this package but someone might think so just by the word async, I would have I think.

That's very true 👍 💯 

> we maybe name the package annotationsasyncsupport or something similar to tie it closer to the instrumentation?

Done.

You've completely convinced me about the split (it'd make instrumentation-api much cleaner!), now I'm starting to think if we have any other scenarios like this one that can benefit from it 😄